### PR TITLE
PP-5302 Modify PaymentViewReponse to be generic for re-use.

### PR DESCRIPTION
**Pull request by danworth**
- Rename `PaymentViewResponse` to `SearchResponse` and make it generic to permit
parameterisation of the contents of the `List` of results. This means it can be
reused for Mandates as well as Payments.
- Move into the common package.
- Refactor `PaymentSearchService` for clarity
- Rename `ViewPaginationBuilder` to `LinksForSearchResult` to hopefully better
explain what it does, also remove the notion of it being a builder because it isn't
and make as many fields `final` as possible.
- Note that within existing logic for `LinksForSearchResult.buildLinks()` it mutates
the page number within the `searchParams` it was provided to create the links
which is a little confusing. In future PR we hope to change this. This is why
`long originalPageNumber = validatedSearchParams.getPage();` has been added for now
within `PaymentSearchService.getPaymentSearchResponse()`.

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/common/model/SearchResponse.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        SearchResponse that = (SearchResponse) o;
+        return Objects.equals(gatewayExternalId, that.gatewayExternalId) &&
+                Objects.equals(total, that.total) &&
+                Objects.equals(page, that.page);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(gatewayExternalId, total, page);
+    }
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/common/model/SearchResponse.java]: Should all the fields be in the `equals(…)` and `hashCode()` methods?

File: src/main/java/uk/gov/pay/directdebit/payments/model/LinksForSearchResult.java [hunk 2]
```diff
@@ -12,20 +12,20 @@
 import static com.fasterxml.jackson.annotation.JsonInclude.Include;
 
 @JsonInclude(Include.NON_NULL)
-public class ViewPaginationBuilder {
+public class LinksForSearchResult {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/payments/model/LinksForSearchResult.java]: This is a much better name

File: src/main/java/uk/gov/pay/directdebit/payments/services/PaymentSearchService.java [hunk 3]
```diff
@@ -22,31 +21,26 @@
...
         this.gatewayAccountDao = gatewayAccountDao;
     }
 
-    public PaymentViewResponse getPaymentSearchResponse(PaymentViewSearchParams searchParams) {
-        return gatewayAccountDao.findByExternalId(searchParams.getGatewayExternalId())
-                .map(gatewayAccount -> {
-                    PaymentViewSearchParams validatedSearchParams = paymentViewValidator.validateParams(searchParams);
-                    List<PaymentResponse> viewListResponse = Collections.emptyList();
-                    Long total = getTotal(validatedSearchParams);
-                    if (total > 0) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/payments/services/PaymentSearchService.java]: Why was this condition removed? Optimisation just not worth it?

File: src/main/java/uk/gov/pay/directdebit/common/model/SearchResponse.java (hunk 1 above)
Comment (danworth) [reply src/main/java/uk/gov/pay/directdebit/common/model/SearchResponse.java]: Fair point, I've regened the methods.

File: src/main/java/uk/gov/pay/directdebit/payments/services/PaymentSearchService.java (hunk 3 above)
Comment (danworth) [reply src/main/java/uk/gov/pay/directdebit/payments/services/PaymentSearchService.java]: Going to add it back.

**Pull request merged**